mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,424 @@
+                    }
+                }
+
+                message = wsocket.next().fuse() => {
+                    match self.handle_response(message, &mut response_notifiers).await {
+                        OuterAction::None => {},
+                        OuterAction::Continue => continue,
+                        OuterAction::Break => break,
+                        OuterAction::Return => return,
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn stop_connection_loop(&self) {
+        let mut tx = self.controller_channel.tx.lock().await;
+        tx.send(ControllerMessage::Close)
+            .await
+            .expect("receiver channel must be alive");
+
+        let response_map = unsafe { &mut *self.responses.0 };
+        response_map.clear();
+    }
+
+    pub(crate) fn maybe_spawn_connection_loop(&self, coin: EthCoin) {
+        self.maybe_spawn_connection_loop_inner(coin, None)
+    }
+
+    pub(crate) fn maybe_spawn_temporary_connection_loop(&self, coin: EthCoin, expires_at: Instant) {
+        self.maybe_spawn_connection_loop_inner(coin, Some(expires_at))
+    }
+
+    fn maybe_spawn_connection_loop_inner(&self, coin: EthCoin, expires_at: Option<Instant>) {
+        // if we can acquire the lock here, it means connection loop is not alive
+        if self.connection_guard.try_lock().is_some() {
+            let fut = self.clone().start_connection_loop(expires_at);
+            let settings = AbortSettings::info_on_abort(format!("connection loop stopped for {:?}", self.node.uri));
+            coin.spawner().spawn_with_settings(fut, settings);
+        }
+    }
+}
+
+async fn send_request(
+    transport: WebsocketTransport,
+    request: Call,
+    request_id: RequestId,
+    event_handlers: Vec<RpcTransportEventHandlerShared>,
+) -> Result<serde_json::Value, Error> {
+    let mut serialized_request = to_string(&request);
+
+    if transport.node.gui_auth {
+        match handle_gui_auth_payload(&transport.gui_auth_validation_generator, &request) {
+            Ok(r) => serialized_request = r,
+            Err(e) => {
+                return Err(Error::Transport(TransportError::Message(format!(
+                    "Couldn't generate signed message payload for {:?}. Error: {e}",
+                    request
+                ))));
+            },
+        };
+    }
+
+    let mut tx = transport.controller_channel.tx.lock().await;
+
+    let (notification_sender, notification_receiver) = futures::channel::oneshot::channel::<()>();
+
+    event_handlers.on_outgoing_request(serialized_request.as_bytes());
+
+    tx.send(ControllerMessage::Request(WsRequest {
+        request_id,
+        serialized_request,
+        response_notifier: notification_sender,
+    }))
+    .await
+    .map_err(|e| Error::Transport(TransportError::Message(e.to_string())))?;
+
+    if let Ok(_ping) = notification_receiver.await {
+        let response_map = unsafe { &mut *transport.responses.0 };
```

Comment:
will never run along this one?

Reply:
> even with a different key

What's the problem of trying to modify different keys at the same time?

Reply:
I don't really know :/, I will need to check that out

I think about it this way: Hashmap is in itself a struct that has some sort of state in it. It could be an array of linked lists, avl tree, red-black tree, etc... so if writing to different keys from each thread might respect each item's own single writer rule, this might not respect the hashmap itself. i don't really have a concrete argument here about the fact but generally i wouldn't run some mutable methods (`.remove()`/`.insert()`) from different threads at the same time if they are not labeled explicitly as thread-safe.

Reply:
I get the point but we don't do updates on existing values.

What we do is:

- Write atomic and unique ID to the map
- Notify other thread
- Remove ID from the map

There can't be data race in this flow.

Reply:
Should be resolved by https://github.com/KomodoPlatform/komodo-defi-framework/pull/2058/commits/9de616c630fd946ae2b9e300dcd29f5c52105b1a